### PR TITLE
[LI-HOTFIX] Reject metadata updates when cluster id changes

### PR DESCRIPTION
**Pull request by kehuum**
Problem:
When brokers of one cluster are removed to another cluster, we noticed that the consumers will start consuming from the new cluster, since upon metadata update response, there is no validation for the cluster id and client just updates the cluster to the cluster in the metadata response.

Fix:
Since such operation like moving brokers between clusters is common, it's better to just reject the metadata update response if cluster id changes and let later updates from the same cluster can proceed.

### CODE REVIEW

**Review by radai-rosenblatt — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 1]
```diff
@@ -252,6 +252,26 @@ public synchronized void update(Cluster newCluster, Set<String> unavailableTopic
...
+            // On client side, the cluster id in Metadata is only null during bootstrap, and client is
+            // expected to talk to the same cluster during its life cycle. Therefore if cluster id changes
+            // during metadata update, meaning this metadata update response is from a different cluster,
+            // client should reject this response and not update cached cluster to the wrong cluster
+
+            // According to SRE, removing brokers and adding to another cluster is common operation, thus
+            // it might be more suitable to just throw an exception for update and fail this update operation
+            // instead of bringing down the client completely, so that the metadata can be updated later from
+            // other brokers in the same cluster.
+            // Related issue: LIKAFKA-32543 since KCA uses regex for topics and has wildcard ACLs, it's more likely
```
Comment (jonlee2) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Since this is public, we probably should not mention LI-specific in the comment.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 2]
```diff
@@ -252,6 +252,33 @@ public synchronized void update(Cluster newCluster, Set<String> unavailableTopic
         if (isClosed())
             throw new IllegalStateException("Update requested after metadata close");
 
+        String previousClusterId = cluster.clusterResource().clusterId();
```
Comment (radai-rosenblatt) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: is it possible for cluster to be null? on 1st connection?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 3]
```diff
@@ -252,6 +252,33 @@ public synchronized void update(Cluster newCluster, Set<String> unavailableTopic
...
+        String newClusterId = newCluster.clusterResource().clusterId();
+
+        if (previousClusterId != null && !newClusterId.equals(previousClusterId)) {
+            // kafka cluster id is unique.
+            // On client side, the cluster id in Metadata is only null during bootstrap, and client is
+            // expected to talk to the same cluster during its life cycle. Therefore if cluster id changes
+            // during metadata update, meaning this metadata update response is from a different cluster,
+            // client should reject this response and not update cached cluster to the wrong cluster
+
+            // According to SRE, removing brokers and adding to another cluster is common operation, thus
```
Comment (radai-rosenblatt) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: this paragraph may be too linkedin specific ?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 4]
```diff
@@ -252,6 +252,33 @@ public synchronized void update(Cluster newCluster, Set<String> unavailableTopic
...
+            // it might be more suitable to just throw an exception for update and fail this update operation
+            // instead of bringing down the client completely, so that the metadata can be updated later from
+            // other brokers in the same cluster.
+
+            // remove the traitor broker from the cluster's node list
+            List<Node> originalNodes = new ArrayList<>(cluster.nodes());
+            for (Node node : newCluster.nodes()) {
+                originalNodes.remove(node);
```
Comment (radai-rosenblatt) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: do we want to count the hits? we expect at least 1 ...

Comment (radai-rosenblatt) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: do we also want to make sure we are left with >0 still-valid brokers ?

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 5]
```diff
@@ -522,7 +523,24 @@ private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long
         long updatedNow = this.time.milliseconds();
         List<ClientResponse> responses = new ArrayList<>();
         handleCompletedSends(responses, updatedNow);
-        handleCompletedReceives(responses, updatedNow);
+
+        try {
+            handleCompletedReceives(responses, updatedNow);
+        } catch (StaleMetadataException e) {
+            // close the connections to the traitor brokers upon this StateMetadataException
+            List<String> currentNodes = new ArrayList<>();
```
Comment (radai-rosenblatt) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: we want to log how many broker connections we closed, and how many valid brokers we are left with?

File: clients/src/main/java/org/apache/kafka/clients/ClientUtils.java [hunk 6]
```diff
@@ -50,15 +52,20 @@ private ClientUtils() {}
...
                         throw new ConfigException("Invalid url in " + CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG + ": " + url);
 
-                    InetSocketAddress address = new InetSocketAddress(host, port);
-
-                    if (address.isUnresolved()) {
-                        log.warn("Removing server {} from {} as DNS resolution failed for {}", url, CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, host);
-                    } else {
-                        addresses.add(address);
-                    }
+                        InetAddress[] inetAddresses = InetAddress.getAllByName(host);
```
Comment (radai-rosenblatt) [inline clients/src/main/java/org/apache/kafka/clients/ClientUtils.java]: is this part of the fix for brokers switching clusters, or a backport of another fix? if its a backport, do we want to make it a different PR ?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 2 above)
Comment (kehuum) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: No, the clusterId will be null on 1st connection, but the cluster object itself will always be non-null.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 3 above)
Comment (kehuum) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Sure, will rephrase this.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 4 above)
Comment (kehuum) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: sure

File: clients/src/main/java/org/apache/kafka/clients/ClientUtils.java (hunk 6 above)
Comment (kehuum) [reply clients/src/main/java/org/apache/kafka/clients/ClientUtils.java]: this is a backport of 2.4 client fix, sure, will open another PR for this.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 7]
```diff
@@ -522,7 +523,30 @@ private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long
         long updatedNow = this.time.milliseconds();
         List<ClientResponse> responses = new ArrayList<>();
         handleCompletedSends(responses, updatedNow);
-        handleCompletedReceives(responses, updatedNow);
+
+        try {
+            handleCompletedReceives(responses, updatedNow);
+        } catch (StaleMetadataException e) {
+            // close the connections to the traitor brokers upon this StateMetadataException
```
Comment (radai-rosenblatt) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: nit - typo in StateMetadata...

**Review by radai-rosenblatt — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 8]
```diff
@@ -285,9 +285,10 @@ public synchronized void update(Cluster newCluster, Set<String> unavailableTopic
             this.cluster = newCluster;
         }
 
+        String previousClusterId = this.cluster.clusterResource().clusterId();
```
Comment (xiowu0) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: move inside if statement?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 9]
```diff
@@ -397,6 +398,47 @@ private synchronized void requestUpdateForNewTopics() {
...
+
+            // Since removing brokers and adding to another cluster can be common operation,
+            // it might be more suitable to just throw an exception for update and fail this update operation
+            // instead of bringing down the client completely, so that the metadata can be updated later from
+            // other brokers in the same cluster.
+
+            // remove the traitor broker from the cluster's node list
+            List<Node> originalNodes = new ArrayList<>(this.cluster.nodes());
+            for (Node node : newCluster.nodes()) {
+                originalNodes.remove(node);
```
Comment (xiowu0) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: The node hash code is computed on (id, host, port, rack).  If the node is moved to a different rack in a different cluster, this doesn't clean up stale node?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 10]
```diff
@@ -397,6 +398,47 @@ private synchronized void requestUpdateForNewTopics() {
...
+            }
+
+            // 1. If all brokers in current cluster has been removed, close connections to all the brokers since
+            //    they are not valid anymore
+            // 2. If there are still active brokers in current cluster, only close connections to those removed
+            //    brokers
+            if (originalNodes.size() == 0) {
+                log.error("No valid brokers in current cluster {} anymore, please reboot the producer/consumer!", previousClusterId);
```
Comment (xiowu0) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: There isn't a recovery path from here, should we try to close the client in this case?

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 11]
```diff
@@ -522,7 +523,30 @@ private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long
...
+
+        try {
+            handleCompletedReceives(responses, updatedNow);
+        } catch (StaleMetadataException e) {
+            // close the connections to the traitor brokers upon this StaleMetadataException
+            List<String> currentNodes = new ArrayList<>();
+            for (Node node : this.metadataUpdater.fetchNodes()) {
+                currentNodes.add(node.idString());
+            }
+
```
Comment (xiowu0) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: should we follow the procedure in "processDisconnection" to clean up all inflight requests that are targeted to these nodes?

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 12]
```diff
@@ -522,7 +522,16 @@ private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long
         long updatedNow = this.time.milliseconds();
         List<ClientResponse> responses = new ArrayList<>();
         handleCompletedSends(responses, updatedNow);
-        handleCompletedReceives(responses, updatedNow);
+
+        try {
+            handleCompletedReceives(responses, updatedNow);
+        } catch (StaleMetadataException e) {
```
Comment (xiowu0) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: is "StaleMetadataException" thrown only in validateCluster? is there any other case that could end up with StaleMetadataException?  If so, maybe we should have a new type of exception. Can you add a unit test for it?

**Review by xiowu0 — APPROVED** [review]
